debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/ehcache/EhcacheLogMinerTransactionCache.java

```diff
@@ -0,0 +1,187 @@
+    private final Cache<String, LogMinerEvent> eventCache;
+
+    // Heap-backed caches for quick access to specific metadata to speed up processing
+    private final Map<String, TreeSet<Integer>> eventIdsByTransactionId = new HashMap<>();
+
+    public EhcacheLogMinerTransactionCache(Cache<String, EhcacheTransaction> transactionCache, Cache<String, LogMinerEvent> eventCache) {
+        this.transactionCache = transactionCache;
+        this.eventCache = eventCache;
+    }
+
+    @Override
+    public EhcacheTransaction getTransaction(String transactionId) {
+        return transactionCache.get(transactionId);
+    }
+
+    @Override
+    public void addTransaction(EhcacheTransaction transaction) {
+        transactionCache.put(transaction.getTransactionId(), transaction);
+        eventIdsByTransactionId.put(transaction.getTransactionId(), new TreeSet<>());
+    }
+
+    @Override
+    public void removeTransaction(EhcacheTransaction transaction) {
+        transactionCache.remove(transaction.getTransactionId());
+    }
+
+    @Override
+    public boolean containsTransaction(String transactionId) {
+        return eventIdsByTransactionId.containsKey(transactionId);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return eventIdsByTransactionId.isEmpty();
+    }
+
+    @Override
+    public int getTransactionCount() {
+        return eventIdsByTransactionId.size();
+    }
+
+    @Override
+    public <R> R streamTransactionsAndReturn(Function<Stream<EhcacheTransaction>, R> consumer) {
+        try (var stream = StreamSupport.stream(transactionCache.spliterator(), false)) {
+            return consumer.apply(stream.map(Cache.Entry::getValue));
+        }
+    }
+
+    @Override
+    public void transactions(Consumer<Stream<EhcacheTransaction>> consumer) {
+        try (var stream = StreamSupport.stream(transactionCache.spliterator(), false)) {
+            consumer.accept(stream.map(Cache.Entry::getValue));
+        }
+    }
+
+    @Override
+    public void eventKeys(Consumer<Stream<String>> consumer) {
+        try (var stream = StreamSupport.stream(eventCache.spliterator(), false)) {
+            consumer.accept(stream.map(Cache.Entry::getKey));
+        }
+    }
+
+    @Override
+    public Iterator<LogMinerEvent> eventsIterator(EhcacheTransaction transaction) {
+        return eventIdsByTransactionId.get(transaction.getTransactionId())
+                .stream()
+                .map(eventId -> getTransactionEvent(transaction, eventId))
+                .iterator();
+    }
+
+    @Override
+    public LogMinerEvent getTransactionEvent(EhcacheTransaction transaction, int eventKey) {
+        return eventCache.get(transaction.getEventId(eventKey));
+    }
+
+    @Override
+    public EhcacheTransaction getAndRemoveTransaction(String transactionId) {
+        final EhcacheTransaction transaction = getTransaction(transactionId);
+        if (transaction != null) {
+            transactionCache.remove(transactionId);
```

Comment:
eventIdsByTransactionId.remove

Reply:
So this is a really poor side effect of how we manage the transaction/events during the commit handler.  We eject the transaction, process the events, and then eject the events.  I think ideally I'd prefer we not eject anything from the cache until the end, and I think if we do that when `removeTransaction` and `getAndRemoveTransaction` and simply cleaer the transaction & events.  :+1:

Reply:
So I see why we do it the way we currently do. The idea is we rely on the transaction cache after we remove the transaction to calculate the smallest scn based on the remaining transactions. I suppose we could leave the transaction in and simply have that transaction excluded from the computation as an alternative :+1: